src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,887 @@
+    ssize_t coincident = 0;
+
+    for (ssize_t i=0; i<genome1.NumLink(); ++i) {
+      int linkEnabledInGenome1 = (int) genome1.aLinkGenes[i].Enabled();
+      ssize_t innovId = genome1.aLinkGenes[i].InnovationId();
+      ssize_t idx = genome2.GetLinkIndex(innovId);
+      bool linkContainedInGenome2 = (idx != -1);
+
+      if (linkContainedInGenome2) {
+        int linkEnabledInGenome2 = (int) genome2.aLinkGenes[idx].Enabled();
+        deltaW += std::abs(genome1.aLinkGenes[i].Weight() * linkEnabledInGenome1 - 
+                           genome2.aLinkGenes[idx].Weight() * linkEnabledInGenome2);
+        ++coincident;
+      }
+    }
+
+    deltaW = deltaW / coincident;
+    return deltaW;
+  }
+
+  // Whether two genome belong to same species or not.
+  bool IsSameSpecies(Genome& genome1, Genome& genome2) {
+    double deltaD = Disjoint(genome1, genome2);
+    double deltaW = WeightDiff(genome1, genome2);
+    double delta = aCoeffDisjoint * deltaD + aCoeffWeightDiff * deltaW;
+
+    if (delta < aCompatThreshold) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // Add genome to existing species or create new species.
+  void AddGenomeToSpecies(Population& population, Genome& genome) {
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      if (population.aSpecies[i].aGenomes.size() > 0) {
+        if (IsSameSpecies(population.aSpecies[i].aGenomes[0], genome)) {  // each first genome in species is the representative genome.
+          population.aSpecies[i].AddGenome(genome);
+          return;
+        }
+      }
+    }
+
+    Species newSpecies = Species();
+    newSpecies.AddGenome(genome);
+    newSpecies.StaleAge(0);
+    population.AddSpecies(newSpecies);
+  }
+
+  // Remove stale species.
+  void RemoveStaleSpecies(Population& population) {
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      if (population.aSpecies[i].StaleAge() > aStaleAgeThreshold) {
+        population.RemoveSpecies(i);
+      }
+    }
+  }
+
+  // Aggregate population's genomes.
+  void AggregateGenomes(Population& population, std::vector<Genome>& genomes) {
+    genomes.clear();
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      for (ssize_t j=0; j<population.aSpecies[i].aGenomes.size(); ++j) {
+        genomes.push_back(population.aSpecies[i].aGenomes[j]);
+      }
+    }
+  }
+
+  // Sort genomes by fitness. Smaller fitness is better and put first.
+  void SortGenomes(std::vector<Genome>& genomes) {
+    std::sort(genomes.begin(), genomes.end(), Species::CompareGenome);
+  }
+
+  // Get genome index in a genomes vector.
+  ssize_t GetGenomeIndex(std::vector<Genome>& genomes, ssize_t id) {
+    for (ssize_t i=0; i<genomes.size(); ++i) {
+      if (genomes[i].Id() == id)
+        return i;
+    }
+    return -1;
+  }
+
+  // Calculate species' average rank in population by fitness. Bigger is better.
+  void CalcSpeciesAverageRank(Population& population, std::vector<double>& speciesAverageRank) {
```

Comment:
I am not sure, but after we sort with fitness, isn't the position is
exactly the rank of genomes?

2016-07-17 15:25 GMT-06:00 Marcus Edel <notifications@github.com>:

> In src/mlpack/methods/ne/neat.hpp
> <https://github.com/mlpack/mlpack/pull/686#discussion_r71089856>:
>
> > +  // Sort genomes by fitness. Smaller fitness is better and put first.
> > +  void SortGenomes(std::vector<Genome>& genomes) {
> > +    std::sort(genomes.begin(), genomes.end(), Species::CompareGenome);
> > +  }
> > +
> > +  // Get genome index in a genomes vector.
> > +  ssize_t GetGenomeIndex(std::vector<Genome>& genomes, ssize_t id) {
> > +    for (ssize_t i=0; i<genomes.size(); ++i) {
> > +      if (genomes[i].Id() == id)
> > +        return i;
> > +    }
> > +    return -1;
> > +  }
> > +
> > +  // Calculate species' average rank in population by fitness. Bigger is better.
> > +  void CalcSpeciesAverageRank(Population& population, std::vector<double>& speciesAverageRank) {
>
> I thought it would be easier to rank the genomes, directly by the fitness
> value instead of checking at which position the genome is after sorting.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mlpack/mlpack/pull/686/files/fc982b9406838eb9697000481e816165702e8ba9#r71089856>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD9tIW74f1sPpqVfwNsFlmqZzVJ1A1Niks5qWp3VgaJpZM4IwJa6>
> .
>